src/Network/Minio/Data.hs

```diff
@@ -185,102 +201,85 @@ type PartTuple = (PartNumber, ETag)
 -- | Represents result from a listing of object parts of an ongoing
 -- multipart upload.
 data ListPartsResult = ListPartsResult {
-    lprHasMore :: Bool
+    lprHasMore  :: Bool
   , lprNextPart :: Maybe Int
-  , lprParts :: [ObjectPartInfo]
+  , lprParts    :: [ObjectPartInfo]
  } deriving (Show, Eq)
 
 
 -- | Represents information about an object part in an ongoing
 -- multipart upload.
 data ObjectPartInfo = ObjectPartInfo {
-    opiNumber :: PartNumber
-  , opiETag :: ETag
-  , opiSize :: Int64
+    opiNumber  :: PartNumber
+  , opiETag    :: ETag
+  , opiSize    :: Int64
   , opiModTime :: UTCTime
   } deriving (Show, Eq)
 
 -- | Represents result from a listing of incomplete uploads to a
 -- bucket.
 data ListUploadsResult = ListUploadsResult {
-    lurHasMore :: Bool
-  , lurNextKey :: Maybe Text
+    lurHasMore    :: Bool
+  , lurNextKey    :: Maybe Text
   , lurNextUpload :: Maybe Text
-  , lurUploads :: [(Object, UploadId, UTCTime)]
-  , lurCPrefixes :: [Text]
+  , lurUploads    :: [(Object, UploadId, UTCTime)]
+  , lurCPrefixes  :: [Text]
   } deriving (Show, Eq)
 
 -- | Represents information about a multipart upload.
 data UploadInfo = UploadInfo {
-    uiKey :: Object
+    uiKey      :: Object
   , uiUploadId :: UploadId
   , uiInitTime :: UTCTime
-  , uiSize :: Int64
+  , uiSize     :: Int64
   } deriving (Show, Eq)
 
 -- | Represents result from a listing of objects in a bucket.
 data ListObjectsResult = ListObjectsResult {
-    lorHasMore :: Bool
+    lorHasMore   :: Bool
   , lorNextToken :: Maybe Text
-  , lorObjects :: [ObjectInfo]
+  , lorObjects   :: [ObjectInfo]
   , lorCPrefixes :: [Text]
   } deriving (Show, Eq)
 
 -- | Represents result from a listing of objects version 1 in a bucket.
 data ListObjectsV1Result = ListObjectsV1Result {
-    lorHasMore' :: Bool
+    lorHasMore'   :: Bool
   , lorNextMarker :: Maybe Text
-  , lorObjects' :: [ObjectInfo]
+  , lorObjects'   :: [ObjectInfo]
   , lorCPrefixes' :: [Text]
   } deriving (Show, Eq)
 
 -- | Represents information about an object.
 data ObjectInfo = ObjectInfo {
-    oiObject :: Object
+    oiObject  :: Object
   , oiModTime :: UTCTime
-  , oiETag :: ETag
-  , oiSize :: Int64
+  , oiETag    :: ETag
+  , oiSize    :: Int64
   } deriving (Show, Eq)
 
-data CopyPartSource = CopyPartSource {
-    -- | formatted like "\/sourceBucket\/sourceObject"
-    cpSource :: Text
-    -- | (0, 9) means first ten bytes of the source object
-  , cpSourceRange :: Maybe (Int64, Int64)
-  , cpSourceIfMatch :: Maybe Text
-  , cpSourceIfNoneMatch :: Maybe Text
-  , cpSourceIfUnmodifiedSince :: Maybe UTCTime
-  , cpSourceIfModifiedSince :: Maybe UTCTime
+-- | Represents source object in server-side copy object
+data SourceInfo = SourceInfo {
+    srcBucket            :: Text
+  , srcObject            :: Text
+  , srcRange             :: Maybe HT.ByteRange
```

Comment:
`ByteRange` does not seem appropriate here. In the basic CopyObjectPart API, both start and end are required to be specified. In the higher-level CopyObject/ComposeObject APIs we can work with specifying a prefix/suffix of the object just like in HTTP range headers (since we would be first performing a HEAD request to get the size of the source object), however I don't think this is significantly useful. I think it is ok to require both start and end for this API.

Reply:
Agreed. We can have `srcRange :: Maybe (Int64, Int64)` instead of `Maybe HT.ByteRange`.